### DISCUSSION TITLE
Very simple model of water gravity flow between two vessels fails when vessel is overfilled. How to add constraints to the model?

### DISCUSSION THREAD

**Discussion by shevart75** [OpenModelica/OpenModelica#11090] [Q&A] [answered]
a very simple model, 4 variables vs. 4 equations! for water flow from vessel 1 to vessel 2 works quite well. But when vessel 2 gets full, the flow should stop and the pressure in vessel 2 shall become equal to that in vessel 1 (physics!). If I add this constraint to the model, the simulation fails at exactly this time point.

See attached model TestOverflow
The code that fails:
if p2<HMax2 * rho * g then  // normal flow - WORKS!!!
  Q1=A * der(p1) / g; // flow into vessel 1 increases the level = pressure/rho/g
  Q2=A * der(p2) / g;
else // vessel 2 is full, no flow, pressure p2 shall then become equal to p1!!! PROBLEM HERE!!! HOW TO REACH THIS???
  p2=p1;
  Q1=0;
end if;

Physically expect that if vessel 2 is full (p2=HMax2rhog) then the simulation shall just continue with a trivial solution of steady state: no flow, pressures are equalized.

So try to set p2 to p1 and flow to 0, if vessel 2 is full: Q1=0; p2=p1;

Without this additional constraint the model works well, except that the vessel 2 gets overfilled. What is wrong with the model?

Thanks a lot!

Error code: Simulation process failed. Exited with code 0xffffffffffffffff. Error solving nonlinear system 8 at time 0.198323 nonlinear system 8 fails: at t=0.198323 [1] Real der(p2)(start=0, nominal=1) 
[TestOverflow_mo.txt](https://github.com/OpenModelica/OpenModelica/files/12447588/TestOverflow_mo.txt)

**Comment by shevart75** [comment, marked as answer]
I'm back with a solution, for those who might need it as well:
when p2>=HMax2 * rho * g then  // the vessel 2 is full, so stop the flow and set pressure same as in vessel 1
  reinit(p2,p1);
end when;

reinit does the job.

**Comment by casella** [comment]
The only problem I see with this mode is that it is not modular. Solving it in a modular way is non-trivial, but it could be done by assuming that when the level gets very close to the maximum the pressure increases a lot more than it does when the level is free to grow. This would correspond to some physical mechanism to goern the behaviour of the tank, e.g. a control valve that lets air out when the level is low enough, but is closed when the level gets close to the maximum level, thus causing an abrupt increase of the pressure of the thin air layer on top of the water level.

**Comment by shevart75** [comment]
Hello,

could you please give an example on how to do it with modelica code:
"to goern the behaviour of the tank, e.g. a control valve that lets air out
when the level is low enough, but is closed when the level gets close to
the maximum level, thus causing an abrupt increase of the pressure of the
thin air layer on top of the water level" ?
At least some main points.

If tank is not full, air pressure on top of water level is 0 (atmospheric).
If tank is full, how to decide to which value the air pressure shall rise?
Shall it rise abruptly to some certain value?
Shall one reinit pressure, or reset the inflow through the tank to zero?
I tried many different things, but none of the solutions was satisfactory.
How to do it with the code?

Thank you!


ср, 6 сент. 2023 г. в 12:18, Francesco Casella ***@***.***>:

**Comment by casella** [comment]
Sure. In order to do that, it's best to use the level instead of the pressure as state variable. Something along these lines
```modelica
rho*A*der(l1) = Q1;
p1 = rho*g*l1 + if l1<l_max then 0 else DP*(l1-lmax)
```
with `DP` a sufficiently large constant, but not too much to avoid excessive numerical stiffness. As soon as `l1` goes above `l_max`, `p1` will start to increase rapidly, until it reaches `p2` and the flow stops. This way the module is modular (you can write the models of the two tanks independently, and combine them as you wish, using connectors) and has a physical motivation, as explained above.

In fact, assuming that once the valve is closed the amount of air above the level is fixed and at constant temperature, you could make it even more physical, e.g.:

```modelica
rho*A*der(l1) = Q1;
p1 = rho*g*l1 + p_air - p_atm;
if l1 < l_max then
  p_air = p_atm;
else
  p_air*(V0 - A*(l1-lmax)) = p_atm*V0;
end if;
```
where `V0` is the constant volume occupied by air when the valve closes as the level `l1` reached `l_max`.

In this way, you have a much more declarative way of modelling, which is closer to the physical behaviour. The second formulation has a parameter `V0` which is easier to give a reasonable numerical value to, compared to the `DP` of the first model.

Your proposed solution is more an algorithmic one, and as such it is not modular and cannot be readily reused in other contexts (e.g. more than two tanks).

Hope this helps 😅

**Comment by casella** [comment]
Note the beauty of using a fully declarative language as Modelica. You write equations that make physical sense, and then let the tool worry about sorting and solving them appropriately.
